fix GEM inputs for `L1REPACK:FullMC`

missirol opened the pull request:
#### PR description:

It looks like `L1REPACK:FullMC` does not work as expected on EDM inputs without the DIGI data tier, as [1] returns [2].

(This would suggest that, up to now, users were simply reading the GEM digis from the EDM inputs when running `L1REPACK:FullMC`.)

This PR is a tentative fix (I'm not a domain expert). It is based on what is currently done for `L1REPACK:Full` (which is for real data, as opposed to `L1REPACK:FullMC` which is for MC), e.g.
https://github.com/cms-sw/cmssw/blob/a8e4ccae45c0e50fba6f716327e0ef4bfd1f145f/Configuration/StandardSequences/python/SimL1EmulatorRepack_Full_cff.py#L168-L170
(`L1REPACK:Full` was fixed in #36133, based on https://github.com/cms-sw/cmssw/pull/36133#issuecomment-971473027)

[1]
```
cmsDriver.py step2 -s L1REPACK:FullMC --conditions auto:phase1_2022_realistic_postEE --era Run3 \
  --datatier RAW --eventcontent RAW --geometry DB:Extended --nThreads 1 --nStreams 0 -n 10 \
  --filein /store/mc/Run3Summer22EEDR/QCD_Pt-120To170_TuneCP5_13p6TeV-pythia8/GEN-SIM-RAW/Poisson70KeepRAW_124X_mcRun3_2022_realistic_postEE_v1-v1/80000/55194757-7f7a-4fac-83b2-fc6e0b5fe7f5.root
```

[2]
```
----- Begin Fatal Exception 09-Jan-2023 18:08:53 CET-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 104 event: 103002 stream: 0
   [1] Running path 'RAWoutput_step'
   [2] Prefetching for module PoolOutputModule/'RAWoutput'
   [3] Prefetching for module RawDataCollectorByLabel/'rawDataCollector'
   [4] Prefetching for module L1TDigiToRaw/'packGmtStage2'
   [5] Prefetching for module L1TMuonOverlapPhase1TrackProducer/'simOmtfDigis'
   [6] Prefetching for module CSCTriggerPrimitivesProducer/'simCscTriggerPrimitiveDigis'
   [7] Prefetching for module GEMPadDigiClusterProducer/'simMuonGEMPadDigiClusters'
   [8] Calling method for module GEMPadDigiProducer/'simMuonGEMPadDigis'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: MuonDigiCollection<GEMDetId,GEMDigi>
Looking for module label: simMuonGEMDigis
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```

#### PR validation:

Nothing beyond the use case described in the PR description.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

~~Plan to backport to (at least) `12_6_X` to aid HLT studies.~~

`CMSSW_12_4_X`
`CMSSW_12_5_X`
`CMSSW_12_6_X`

missirol committed:
fix GEM inputs for L1REPACK:FullMC

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40461/33624

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @missirol (Marino Missiroli) for master.

It involves the following packages:

- Configuration/StandardSequences (**operations**)


@cmsbuild, @perrotta, @rappoccio, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@fabiocos, @VourMa, @makortel, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @Martin-Grunewald, @lecriste, @mmusich, @sameasy, @mtosi, @dgulhan, @slomeo this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented:
assign l1

cmsbuild commented:
New categories assigned: l1

@epalencia,@rekovic,@cecilecaillol you have been requested to review this Pull request/Issue and eventually sign? Thanks

missirol commented:
please test

@cms-sw/l1-l2 , is there any wf testing `L1REPACK:FullMC` in CMSSW ?

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1dd99c/29854/summary.html
**COMMIT**: e290d429c270dae5dc61c0c2db06d311f3c18fb5
**CMSSW**: CMSSW_13_0_X_2023-01-09-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40461/29854/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 12 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 160
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555356
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-09-1100+1dd99c/54872/triggerResults): no differences found

missirol commented:
@cms-sw/l1-l2 

Could you please have a look, and check if this fix is correct? If so, it would be useful to have this in `13_0_0_pre3` (deadline: today), and it will have to be backported. Thanks!

cecilecaillol commented:
+l1

perrotta commented on Configuration/StandardSequences/python/SimL1EmulatorRepack_FullMC_cff.py:
```diff
@@ -60,6 +67,8 @@ def _print(ignored):
 simCscTriggerPrimitiveDigis.CSCComparatorDigiProducer = 'unpackCSC:MuonCSCComparatorDigi'
 simCscTriggerPrimitiveDigis.CSCWireDigiProducer       = 'unpackCSC:MuonCSCWireDigi'
+# GEM
+run3_GEM.toModify(simMuonGEMPadDigis, InputCollection = 'unpackGEM')
```
This is the same as what currently in SimL1EmulatorRepack_Full_cff.py, therefore I imagine it may work and my question is probably silly...

But what happens if `run3_GEM` and not `stage2L1Trigger`? 'unpackGEM' is feed as InputCollection to `simMuonGEMPadDigis` [[1]](https://github.com/cms-sw/cmssw/pull/40461/files#diff-6afbf763bc54eac6d2fb579ea378604e8337d4492446134e43419aae84046af6R71), but the `SimL1EmulatorTask` is not updates with the `unpackGEM` module in it [[2]](https://github.com/cms-sw/cmssw/pull/40461/files#diff-6afbf763bc54eac6d2fb579ea378604e8337d4492446134e43419aae84046af6R129)

missirol replied:
Well, I can only guess as a non-L1T expert.

 - This file only really works for `stage2L1Trigger`, according to the comments up top ("only supports Stage 2 eras for now."), modulo the fact that the comments are mistakenly swapped (the first comment, "use a legacy version", should show for `~stage2L1Trigger`, not for `stage2L1Trigger`)
 - In practise, I think `stage2L1Trigger` applies to nearly anything after 2015, so "`run3_GEM` and not `stage2L1Trigger`" should never happen (in principle..).
 
 Let me fix the comments, since I would like to backport this..

perrotta replied:
Thank you Marino, this is also my understanding.
However, even if so, the clean way to implement I imagine it should be
```suggestion
(stage2L1Trigger & run3_GEM).toModify(simMuonGEMPadDigis, InputCollection = 'unpackGEM')
```

missirol replied:
Following what I see elsewhere, e.g.
https://github.com/cms-sw/cmssw/blob/a45ef0567719756ab6b998f0bd1611151ca695f9/L1Trigger/L1TMuon/python/simDigis_cff.py#L119
I can change `run3_GEM` to `(stage2L1Trigger & run3_GEM)` also above (and in "Full").

missirol replied:
>  However, even if so, the clean way to implement I imagine it should be

I cross-posted. Yes, will do (soon).

missirol commented:
Updated to address https://github.com/cms-sw/cmssw/pull/40461#discussion_r1065857148. Also fixed the mistakenly-swapped comments in other L1REPACK `cff`s. Latest updates tested with [1] (python syntax check).

[1]
```sh
#!/bin/bash

### Check non-stage2L1Trigger case
cat <<@EOF > tmp.py
import sys
import FWCore.ParameterSet.Config as cms
process = cms.Process('L1REPACK')
process.load('Configuration.StandardSequences.SimL1EmulatorRepack_'+sys.argv[1]+'_cff')
@EOF
for foo in CalouGT Full2015Data FullMC FullSimTP Full GCTGT GT1 GT2 GT uGT ; do
  python3 tmp.py $foo
done
unset foo

### Check stage2L1Trigger case
cat <<@EOF > tmp.py
import sys
import FWCore.ParameterSet.Config as cms
from Configuration.Eras.Era_Run3_cff import Run3
process = cms.Process('L1REPACK', Run3)
process.load('Configuration.StandardSequences.SimL1EmulatorRepack_'+sys.argv[1]+'_cff')
@EOF
for foo in CalouGT Full2015Data FullMC FullSimTP Full GCTGT GT1 GT2 GT uGT ; do
  python3 tmp.py $foo
done
unset foo

rm -f tmp.py
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40461/33642

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #40461 was updated. @perrotta, @rappoccio, @epalencia, @cecilecaillol, @cmsbuild, @rekovic, @fabiocos, @davidlange6 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1dd99c/29892/summary.html
**COMMIT**: 633c942bdab6a2318b7fda5bd266180520c0057b
**CMSSW**: CMSSW_13_0_X_2023-01-10-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40461/29892/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 35 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555538
 - DQMHistoTests: Total failures: 1194
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3554322
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-10-1100+1dd99c/54910/triggerResults): no differences found

missirol commented:
@cms-sw/l1-l2, could you please check again, and re-sign? Thanks!

The diff of this PR since your previous signature is below:
https://github.com/cms-sw/cmssw/compare/e290d429c270dae5dc61c0c2db06d311f3c18fb5..633c942bdab6a2318b7fda5bd266180520c0057b

cecilecaillol commented:
+l1

perrotta commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
